dbms/src/TiDB/Schema/TiDB.cpp

```diff
@@ -1362,17 +1398,62 @@ ColumnInfo toTiDBColumnInfo(const tipb::ColumnInfo & tipb_column_info)
     tidb_column_info.flag = tipb_column_info.flag();
     tidb_column_info.flen = tipb_column_info.columnlen();
     tidb_column_info.decimal = tipb_column_info.decimal();
+    tidb_column_info.collate = tipb_column_info.collation();
+    for (int i = 0; i < tipb_column_info.elems_size(); ++i)
+        tidb_column_info.elems.emplace_back(tipb_column_info.elems(i), i + 1);
     // TiFlash get default value from origin_default_value, check `Field ColumnInfo::defaultValueToField() const`
     // So we need to set origin_default_value to tipb_column_info.default_val()
     // Related logic in tidb, https://github.com/pingcap/tidb/blob/45318da24d8e4c0c6aab836d291a33f949dd18bf/pkg/table/tables/tables.go#L2303-L2329
-    // For TypeBit, we need to set origin_default_bit_value to tipb_column_info.default_val().
-    if (tidb_column_info.tp == TypeBit)
-        tidb_column_info.origin_default_bit_value = tipb_column_info.default_val();
-    else
+    if (tidb_column_info.tp != TypeBit)
         tidb_column_info.origin_default_value = tipb_column_info.default_val();
-    tidb_column_info.collate = tipb_column_info.collation();
-    for (int i = 0; i < tipb_column_info.elems_size(); ++i)
-        tidb_column_info.elems.emplace_back(tipb_column_info.elems(i), i + 1);
+    // Decode tipb_column_info.default_val.
+    {
+        // The default value is null.
+        if (tidb_column_info.origin_default_value.size() == 0)
+        {
+            Poco::Dynamic::Var empty_val;
+            tidb_column_info.origin_default_value = empty_val;
+            return tidb_column_info;
+        }
+        size_t cursor = 0;
+        auto val = DB::DecodeDatum(cursor, tipb_column_info.default_val());
+        if (val.getType() == DB::Field::Types::String)
+        {
+            tidb_column_info.origin_default_value = val.get<String>();
+            return tidb_column_info;
+        }
+        switch (tidb_column_info.tp)
+        {
+        case TypeDate:
+        case TypeDatetime:
+        case TypeTimestamp:
+        case TypeSet:
+            tidb_column_info.origin_default_value = val.get<UInt64>();
+            break;
+        case TypeTime:
+            tidb_column_info.origin_default_value = val.get<Int64>();
+            break;
+        case TypeBit:
+            // For TypeBit, we need to set origin_default_bit_value to tipb_column_info.default_val().
+            tidb_column_info.default_bit_value = val.get<UInt64>();
+            break;
+        case TypeYear:
+            // If the year column has 'not null' option and the value is 0, this year value is '0000'.
+            if (val.get<Int64>() == 0)
+            {
+                Poco::Dynamic::Var empty_val;
+                tidb_column_info.origin_default_value = empty_val;
+                break;
+            }
```

Comment:
set `tidb_column_info.origin_default_value = tipb_column_info.default_val();` if `val.get<Int64>() != 0` ?

Reply:
Yes. If we don't do this, after `alter table t add column col2 year not null;`, we get this data is `2000` not `0000`.